subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/DirectDependencyMetadataAdapter.java

```diff
@@ -42,4 +52,42 @@ public void doNotEndorseStrictVersions() {
     public boolean isEndorsingStrictVersions() {
         return getOriginalMetadata().isEndorsingStrictVersions();
     }
+
+    @Override
+    public List<DependencyArtifact> getArtifactSelectors() {
+        return getIvyArtifacts().stream().map(this::asDependencyArtifact).collect(Collectors.toList());
+    }
+
+    private DependencyArtifact asDependencyArtifact(IvyArtifactName ivyArtifactName) {
+        return new DefaultDependencyArtifact(ivyArtifactName.getName(), ivyArtifactName.getType(), ivyArtifactName.getExtension(), ivyArtifactName.getClassifier(), null);
+    }
+
+    private List<IvyArtifactName> getIvyArtifacts() {
+        ModuleDependencyMetadata originalMetadata = getOriginalMetadata();
+        if (originalMetadata instanceof ConfigurationBoundExternalDependencyMetadata) {
```

Comment:
:x: We should also support `GradleDependencyMetadata.getDependencyArtifact()` here for completeness.

It is possible if the [thirdPartyCompatibility](https://github.com/gradle/gradle/blob/master/subprojects/docs/src/docs/design/gradle-module-metadata-latest-specification.md#thirdpartycompatibility-value) feature of GMM is used. I am not sure how our test fixtures support that case.